### PR TITLE
[Container App]: Add `az containerapp ingress access-restriction` group

### PR DESCRIPTION
**Pull request by haroonf**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repository and upgrade the version in the pull request but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
Container App

**Review by StrawnSC — APPROVED** [review]

**Comment by haroonf** [comment]
@zhoxing-ms Could you take a look at this? Want to get this released soon if possible. Thanks

File: src/containerapp/azext_containerapp/_help.py [hunk 1]
```diff
@@ -671,6 +671,38 @@
...
+          az containerapp ingress ip-restriction remove --ip-restriction-name restrictionName
+"""
+
+helps['containerapp ingress ip-restriction show'] = """
+    type: command
+    short-summary: Show ip security restrictions for a container app.
+    examples:
+    - name: Show ip security restrictions.
+      text: |
+          az containerapp ingress ip-restriction show
```
Comment (zhoxing-ms) [inline src/containerapp/azext_containerapp/_help.py:703]: Why does the command in this example have no parameters?

File: src/containerapp/azext_containerapp/commands.py [hunk 2]
```diff
@@ -120,6 +120,11 @@ def load_command_table(self, _):
         g.custom_command('set', 'set_ingress_traffic', exception_handler=ex_handler_factory())
         g.custom_show_command('show', 'show_ingress_traffic')
 
+    with self.command_group('containerapp ingress ip-restriction') as g:
+        g.custom_command('set', 'set_ip_restriction', exception_handler=ex_handler_factory())
+        g.custom_command('remove', 'remove_ip_restriction')
```
Comment (zhoxing-ms) [inline src/containerapp/azext_containerapp/commands.py:125]: Do we need to consider adding `confirmation=True` for the remove command?

File: src/containerapp/azext_containerapp/_params.py [hunk 3]
```diff
@@ -249,6 +249,12 @@ def load_arguments(self, _):
         c.argument('target_port', type=int, validator=validate_target_port, help="The application port used for ingress traffic.")
         c.argument('exposed_port', type=int, help="Additional exposed port. Only supported by tcp transport protocol. Must be unique per environment if the app ingress is external.")
 
+    with self.argument_context('containerapp ingress ip-restriction') as c:
+        c.argument('allow_access', help='Boolean indicating whether the ip security restriction allows or denies access.')
```
Comment (zhoxing-ms) [inline src/containerapp/azext_containerapp/_params.py:253]: Could we define this parameter as `action='store_true'` ?

File: src/containerapp/azext_containerapp/commands.py (hunk 2 above)
Comment (haroonf) [reply src/containerapp/azext_containerapp/commands.py:125]: This is a property, and keeping with the convention in containerapp commands I think it's unnecessary

File: src/containerapp/azext_containerapp/_params.py (hunk 3 above)
Comment (haroonf) [reply src/containerapp/azext_containerapp/_params.py:253]: This parameter changes frequently, and we wouldn't want the default behavior to change (i.e. customers might create an ip restriction that allows when they want to block an ip range)

File: src/containerapp/azext_containerapp/_help.py (hunk 1 above)
Comment (haroonf) [reply src/containerapp/azext_containerapp/_help.py]: fixed

File: src/containerapp/azext_containerapp/_params.py [hunk 4]
```diff
@@ -249,6 +249,15 @@ def load_arguments(self, _):
         c.argument('target_port', type=int, validator=validate_target_port, help="The application port used for ingress traffic.")
         c.argument('exposed_port', type=int, help="Additional exposed port. Only supported by tcp transport protocol. Must be unique per environment if the app ingress is external.")
 
+    with self.argument_context('containerapp ingress access-restriction') as c:
+        c.argument('action', arg_type=get_enum_type(['Allow', 'Deny']), help='Boolean indicating whether the ip security restriction allows or denies access.')
```
Comment (anthonychu) [inline src/containerapp/azext_containerapp/_params.py:253]:
```suggestion
        c.argument('action', arg_type=get_enum_type(['Allow', 'Deny']), help='Whether the IP security restriction allows or denies access. All restrictions must be use the same action. If no restrictions are set, all traffic is allowed.')
```

File: src/containerapp/azext_containerapp/_params.py [hunk 5]
```diff
@@ -249,6 +249,15 @@ def load_arguments(self, _):
...
+        c.argument('rule_name', help="The ip security restriction name.")
```
Comment (anthonychu) [inline src/containerapp/azext_containerapp/_params.py:254]:
```suggestion
        c.argument('rule_name', help="The IP security restriction name.")
```

File: src/containerapp/azext_containerapp/_params.py [hunk 6]
```diff
@@ -249,6 +249,15 @@ def load_arguments(self, _):
...
+        c.argument('description', help="The description of the ip security restriction.")
```
Comment (anthonychu) [inline src/containerapp/azext_containerapp/_params.py:255]:
```suggestion
        c.argument('description', help="The description of the IP security restriction.")
```

File: src/containerapp/azext_containerapp/_params.py [hunk 7]
```diff
@@ -249,6 +249,15 @@ def load_arguments(self, _):
...
+        c.argument('ip_address', help="The ip address range of the ip security restriction.")
```
Comment (anthonychu) [inline src/containerapp/azext_containerapp/_params.py:256]:
```suggestion
        c.argument('ip_address', help="The address range of the IP security restriction in CIDR notation. (for example, '198.51.100.14/24')")
```

File: src/containerapp/azext_containerapp/_help.py [hunk 8]
```diff
@@ -671,6 +671,38 @@
           az containerapp ingress traffic show -n MyContainerapp -g MyResourceGroup
 """
 
+helps['containerapp ingress access-restriction'] = """
+    type: group
+    short-summary: Commands to manage access restrictions.
```
Comment (anthonychu) [inline src/containerapp/azext_containerapp/_help.py:676]:
```suggestion
    short-summary: Commands to manage IP access restrictions.
```

File: src/containerapp/azext_containerapp/_help.py [hunk 9]
```diff
@@ -671,6 +671,38 @@
...
+"""
+
+helps['containerapp ingress access-restriction set'] = """
+    type: command
+    short-summary: Configure access restrictions for a container app.
```
Comment (anthonychu) [inline src/containerapp/azext_containerapp/_help.py:681]:
```suggestion
    short-summary: Configure IP access restrictions for a container app.
```

File: src/containerapp/azext_containerapp/_help.py [hunk 10]
```diff
@@ -671,6 +671,38 @@
...
+    examples:
+    - name: Add an allow access restriction.
```
Comment (anthonychu) [inline src/containerapp/azext_containerapp/_help.py:683]:
```suggestion
    - name: Add an allow IP access restriction.
```

File: src/containerapp/azext_containerapp/_help.py [hunk 11]
```diff
@@ -671,6 +671,38 @@
...
+      text: |
+          az containerapp ingress access-restriction set -n MyContainerapp -g MyResourceGroup --rule-name restrictionName --ip-address 192.168.1.1/32 --description "Restriction description." --action Allow
+"""
+
+helps['containerapp ingress access-restriction remove'] = """
+    type: command
+    short-summary: Remove access restrictions from a container app.
+    examples:
+    - name: Remove an access restriction.
```
Comment (anthonychu) [inline src/containerapp/azext_containerapp/_help.py:692]:
```suggestion
    short-summary: Remove IP access restrictions from a container app.
    examples:
    - name: Remove an IP access restriction.
```

File: src/containerapp/azext_containerapp/_help.py [hunk 12]
```diff
@@ -671,6 +671,38 @@
...
+      text: |
+          az containerapp ingress access-restriction remove -n MyContainerapp -g MyResourceGroup --rule-name restrictionName
+"""
+
+helps['containerapp ingress access-restriction list'] = """
+    type: command
+    short-summary: List access restrictions for a container app.
+    examples:
+    - name: List access restrictions.
```
Comment (anthonychu) [inline src/containerapp/azext_containerapp/_help.py:701]:
```suggestion
    short-summary: List IP access restrictions for a container app.
    examples:
    - name: List IP access restrictions.
```

**Review by anthonychu — APPROVED** [review]

**Comment by haroonf** [comment]
> @zhoxing-ms Could you take a look at this? Want to get this released soon if possible. Thanks

Let me know if there is anything else, thanks!

File: src/containerapp/azext_containerapp/_params.py (hunk 3 above)
Comment (zhoxing-ms) [reply src/containerapp/azext_containerapp/_params.py]: If the design of this parameter is not stable and has not been finalized, please add the `is_preview` flag to it.

File: src/containerapp/azext_containerapp/commands.py (hunk 2 above)
Comment (zhoxing-ms) [reply src/containerapp/azext_containerapp/commands.py:125]: OK

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**